lib/Dialect/TritonGPU/Transforms/WarpSpecialization/PartitionLoops.cpp

```diff
@@ -16,103 +19,331 @@ using namespace mlir;
+                              const WarpSchedule &schedule) {
+  bool ret = false;
+  schedule.iterateOutputs(loop, partition, [&](Operation *op, OpOperand &use) {
+    if (isa<scf::YieldOp>(op) && use.getOperandNumber() == resultIdx &&
+        isa<RankedTensorType>(loop.getResult(resultIdx).getType())) {
+      ret = true;
+    }
+  });
+  return ret;
+}
 
-// Given a loop, erase ops and loop iter args that are not part of the root
-// partition or the provided `partition`.
-static void eraseOtherPartitions(scf::ForOp &loop, const WarpSchedule &schedule,
-                                 const Partition *partition) {
+SmallVector<LoopVarCategory> classifyLoopVars(scf::ForOp loop,
+                                              const Partition *partition,
+                                              const WarpSchedule &schedule) {
   auto inPartition = [&](Operation *op) {
     const Partition *opPartition =
         schedule.getPartition(loop.getBody()->findAncestorOpInBlock(*op));
     return llvm::is_contained({partition, schedule.getRootPartition()},
                               opPartition);
   };
-  llvm::BitVector toErase(loop.getNumRegionIterArgs(), true);
-  for (Operation &op :
-       llvm::make_early_inc_range(loop.getBody()->without_terminator())) {
-    if (!inPartition(&op)) {
-      op.dropAllUses();
-      op.erase();
-      continue;
+  auto isTensorResultFromOtherPartition = [&](int i) {
+    for (auto otherPartition : schedule.getPartitions()) {
+      if (otherPartition.getIndex() == partition->getIndex()) {
+        continue;
+      }
+      if (isTensorResultComputedBy(loop, i, &otherPartition, schedule)) {
+        return true;
+      }
     }
-    // Trace uses into the `scf.yield` to mark the corresponding iter arg as
-    // used.
-    for (OpOperand &use : op.getUses()) {
-      if (use.getOwner() == loop.getBody()->getTerminator())
-        toErase.reset(use.getOperandNumber());
+    return false;
+  };
+
+  SmallVector<LoopVarCategory> categories(loop.getNumRegionIterArgs());
+  for (auto [i, arg] : llvm::enumerate(loop.getRegionIterArgs())) {
+    if (llvm::any_of(arg.getUsers(), inPartition)) {
+      categories[i] = LoopVarCategory::Used;
+    } else if (isTensorResultFromOtherPartition(i)) {
+      categories[i] = LoopVarCategory::TensorResultFromOtherPartition;
+    } else {
+      categories[i] = LoopVarCategory::Unused;
     }
   }
-  for (auto [i, arg, result] :
-       llvm::enumerate(loop.getRegionIterArgs(), loop.getResults())) {
-    if (llvm::any_of(arg.getUsers(), inPartition) || !result.use_empty())
-      toErase.reset(i);
-    else if (toErase.test(i))
-      arg.dropAllUses();
+
+  return categories;
+}
+
+auto getLoopVarIndicesToKeep(
+    scf::ForOp loop, const Partition *partition,
+    const SmallVector<LoopVarCategory> loopVarCategories) {
+  SmallVector<size_t> indices;
+  SmallVector<int> reverseIndices(loop.getNumRegionIterArgs(), -1);
```

Comment:
question: what does -1 signify?

Reply:
It means an invalid index, the corresponding loop variable in the original loop is removed in the new cloned loop.

Reply:
could you use `std::optional<int>` here to be more explicit?